ec2/keypair.py

```diff
@@ -53,29 +78,86 @@ def delete(**kwargs):
-    key_pair_name = ctx.instance.runtime_properties['key_pair_name']
+    key_pair_name = ctx.instance.runtime_properties['aws_resource_id']
     ctx.logger.info('Deleting the keypair.')
 
     try:
         ec2_client.delete_key_pair(key_pair_name)
-    except (EC2ResponseError, BotoServerError) as e:
+    except (boto.exception.EC2ResponseError,
+            boto.exception.BotoServerError) as e:
         raise NonRecoverableError('Error response on key pair delete. '
-                                  'API returned: {0}'.format(e))
+                                  'API returned: {0}'.format(str(e)))
+    finally:
+        ctx.instance.runtime_properties.pop('aws_resource_id', None)
+        ctx.instance.runtime_properties.pop('key_path', None)
+        ctx.logger.debug('Attempted to delete key pair from account.')
 
+    delete_key_file(ctx=ctx)
     ctx.logger.info('Deleted key pair: {0}.'.format(key_pair_name))
 
 
 @operation
-def creation_validation(**kwargs):
-    ec2_client = connection.EC2ConnectionClient().client()
-    ctx.logger.info('Validating that the keypair '
-                    'was created in your account.')
-    key_pair_name = ctx.instance.runtime_properties['key_pair_name']
+def creation_validation(**_):
+    """ This checks that all user supplied info is valid """
+    required_properties = ['resource_id', 'use_external_resource']
+    for property_key in required_properties:
+        utils.validate_node_property(property_key, ctx=ctx)
+
+    key_path = get_key_file_path(ctx=ctx)
+
+    if ctx.node.properties['use_external_resource']:
+        if not search_for_key_file(key_path):
+            raise NonRecoverableError('Use external resource is true, but the '
+                                      'key file does not exist.')
+
+
+def save_key_pair(key_pair_object, ctx):
+    """ Saves the key pair to the file specified in the blueprint. """
+
+    ctx.logger.debug('Attempting to save the key_pair_object.')
 
     try:
-        ec2_client.get_key_pair(key_pair_name)
-    except (EC2ResponseError, BotoServerError) as e:
-        raise NonRecoverableError('Unable to validate that Key Pair exists. '
-                                  'API returned: {0}'.format(e))
+        key_pair_object.save(ctx.node.properties['private_key_path'])
+    except (boto.exception.BotoClientError,
+            OSError) as e:
+        raise NonRecoverableError(str(e))
+
+    key_path = get_key_file_path(ctx=ctx)
+
+    os.chmod(key_path, 0600)
+
+
+def get_key_file_path(ctx):
+    """The key_path is an attribute that gives the full path to the key file.
+    This function creates the path as a string for use by various functions in
+    this module. It doesn't verify whether the path points to anything.
+    """
+
+    path = os.path.expanduser(ctx.node.properties['private_key_path'])
+    key_path = os.path.join(path,
+                            '{0}{1}'.format(
+                                ctx.node.properties['resource_id'],
```

Comment:
check if `aws_resource_id` has already been set in the workflow when this function runs.